feat(transformer/class-properties): support `private_fields_as_properties` assumption

overlookmotel opened the pull request:
Support `private_fields_as_properties` assumption in class properties transform. This assumption is also enabled by the transform's `loose` option.

Optional chain (e.g. `this?.#prop`) is not yet implemented, but all other usages of private fields are supported. We'll handle optional chain in a follow-on PR.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7717?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#7717** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7717?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7717?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#7716** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7716?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7713** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7713?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7712** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7712?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7711** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7711?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7710** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7710?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7709** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7709?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7708** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7708?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7689** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7689?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#7688** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7688?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@Dunqing Can I ask for your advice on this? This PR implements support for `private_fields_as_properties` assumption for everything except optional chain.

Could you possibly look at what Babel produces for optional chain with this "loose" assumption turned on, and advise how difficult it'd be to implement this?

[Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEBiD29oG8BQ1oGIAOAne20AvNAIwDc60EALmDQJbBa31PHQBMlVAtgKY0AFvAAmACgCUKKhmEMIAfgB0OfNkoY5QhSrUEpmrfKXKAZolV4CR7bvOX92Q1QC-qKq0bMvTALKCIhLSaFrQJnq-wLbhOqaYUS5hEQ7wqlExKRZpCXTeSdDu7kA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.26.2&externalPlugins=%40babel%2Fplugin-transform-class-properties%407.25.7%2C%40babel%2Fplugin-transform-class-static-block%407.26.0%2C%40babel%2Fplugin-external-helpers%407.25.9&assumptions=%7B%22privateFieldsAsProperties%22%3Atrue%7D)

We don't *have* to support `private_fields_as_properties`. It's just an optimization (shorter output), rather than required for correctness. Mainly I wanted to implement it now because it makes so many tests pass, which makes it easier to see what we have left.

But if handling optional chaining is going to be a nightmare, maybe we should leave it until later.

Dunqing commented:
### Merge activity

* **Dec 9, 9:27 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
feat(transformer/class-properties): support `private_fields_as_properties` assumption (#7717)

Support `private_fields_as_properties` assumption in class properties transform. This assumption is also enabled by the transform's `loose` option.

Optional chain (e.g. `this?.#prop`) is not yet implemented, but all other usages of private fields are supported. We'll handle optional chain in a follow-on PR.
